experimental/ai-algorithms/NeuralNets/Lib.idr

```diff
@@ -0,0 +1,169 @@
+rawToCleaned (x::xs) = let some_var = removeStopWords x
+                       in some_var ++ rawToCleaned xs
+export
+nn_Layer : List Double -> List Double -> List Double
+nn_Layer [] _ = []
+nn_Layer _ [] = []
+nn_Layer (x::xs) (y::ys) = (x * y + b)::nn_Layer xs ys
+
+--idris code for Softmax activation
+export
+softmax_activation : List Double -> List Double
+softmax_activation xs = let softmax_denom = sum $ map exp $ xs 
+in foldr (\x,acc => (exp x)/softmax_denom::acc) [] xs
+
+--idris code for Cross Entropy loss
+export
+grad_CrossEntropy : List Double -> List Double -> List Double
+grad_CrossEntropy [] _ = []
+grad_CrossEntropy _ [] = []
+grad_CrossEntropy (x::xs) (y::ys) = -(y/x) + (1-y)/(1-x)::grad_CrossEntropy xs ys
+
+-- idris code for Adam's optimizer
+export
+adam_Optimizer : List Double -> List Double -> Double -> Double -> Double -> Double -> Double -> Double -> List Double
+adam_Optimizer [] _ _ _ _ _ _ _ =  []
+adam_Optimizer _ [] _ _ _ _ _ _ =  []
+adam_Optimizer (t::thetas) (gt::gts) alpha m_prev v_prev beta1 beta2 epsilon = do
+              let m_current = (beta1 * m_prev + (1-beta1) * gt)/ (1-beta1)
+              let v_current = (beta2 * v_prev + (1-beta2) *  (db.pow gt 2))/(1-(db.pow beta2 2))
+              let mhat_current = m_current /(1-beta1)
+              let vhat_current = v_current/(1-beta2)
+              (t - alpha * mhat_current / ((sqrt  vhat_current) + epsilon)):: adam_Optimizer thetas gts alpha m_current v_current beta1 beta2 epsilon
+
+
+export
+fullNNLayer : List (List Double) -> List (List Double) -> List (List Double)
+fullNNLayer [] [] = []
+fullNNLayer (x::inputs) (y::weights) = (softmax_activation $ nn_Layer x y)::fullNNLayer inputs weights
+
+
+||| loss_func is a generic function to compute the loss or the cost of all weights learned. Given we have 
+||| any loss function implemented here we can pass it to this loss_func. The purpose of this function is to
+||| provide multiple options for loss functions as opposed to just the Cross entropy loss implemented above
+export
+loss_func : (List Double -> List Double -> List Double ) -> List (List Double) -> List (List Double) -> List (List Double)
+loss_func f [] _ = []
+loss_func f _ [] = []
+loss_func f (p::preds) (t::targets) = f p t :: loss_func f preds targets
+
+
+||| optimizer is a generic function to compute weights. so far only Adam optimzer is implemented
+||| But one can write the SGD RMSprob and other optimizers and then pass them via this function.
+export
+optimizer : ( List Double -> List Double -> Double -> Double -> Double -> Double -> Double -> Double -> List Double) -> List (List Double) -> List (List Double) -> Double -> Double -> Double -> Double -> Double -> Double -> List (List Double)
+optimizer f [] _ _ _ _ _ _ _ = []
+optimizer f _ [] _ _ _ _ _ _ = []
+optimizer f (ws::wss) (cs::costs) alpha  m v beta1 beta2 epsilon =  (f ws cs alpha m v beta1 beta2 epsilon) :: optimizer f wss costs alpha m  v beta1 beta2 epsilon
```

Comment:
Same remark holds here, you'd better pass a record with the parameters for adam_optimizer instead of passing them directly as input arguments.